services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/kafka/RecordHandler.java

```diff
@@ -440,55 +443,55 @@ private void doInstallSwitchRules(final CommandMessage message, String replyToTo
-        List<Long> removedRules = new ArrayList<>();
+
+        ISwitchManager switchManager = context.getSwitchManager();
+
         try {
-            /*
-             * This first part .. we are either deleting one rule, or all non-default rules (the else)
-             */
-            List<Long> toRemove = new ArrayList<>();
-            if (deleteAction == DeleteRulesAction.ONE ) {
-                toRemove.add(request.getOneCookie());
-            } else if (deleteAction == DeleteRulesAction.REMOVE_DROP) {
-                toRemove.add(ISwitchManager.DROP_RULE_COOKIE);
-            } else if (deleteAction == DeleteRulesAction.REMOVE_BROADCAST) {
-                toRemove.add(ISwitchManager.VERIFICATION_BROADCAST_RULE_COOKIE);
-            } else if (deleteAction == DeleteRulesAction.REMOVE_UNICAST) {
-                toRemove.add(ISwitchManager.VERIFICATION_UNICAST_RULE_COOKIE);
-            } else if (deleteAction == DeleteRulesAction.REMOVE_DEFAULTS
-                    || deleteAction == DeleteRulesAction.REMOVE_ADD) {
-                toRemove.add(ISwitchManager.DROP_RULE_COOKIE);
-                toRemove.add(ISwitchManager.VERIFICATION_BROADCAST_RULE_COOKIE);
-                toRemove.add(ISwitchManager.VERIFICATION_UNICAST_RULE_COOKIE);
+            List<Long> removedRules = new ArrayList<>();
+
+            // The case when we either delete by criteria or a specific default rule.
+            DeleteRulesCriteria criteria = request.getCriteria();
+            if(deleteAction != null) {
+                switch (deleteAction) {
+                    case REMOVE_DROP:
+                        criteria = DeleteRulesCriteria.builder()
+                                .cookie(ISwitchManager.DROP_RULE_COOKIE).build();
+                        break;
+                    case REMOVE_BROADCAST:
+                        criteria = DeleteRulesCriteria.builder()
+                                .cookie(ISwitchManager.VERIFICATION_BROADCAST_RULE_COOKIE).build();
+                        break;
+                    case REMOVE_UNICAST:
+                        criteria = DeleteRulesCriteria.builder()
+                                .cookie(ISwitchManager.VERIFICATION_UNICAST_RULE_COOKIE).build();
+                        break;
+                }
+            }
+            if (criteria != null) {
+                removedRules.addAll(switchManager.deleteRulesByCriteria(dpid, criteria));
             }
 
-            // toRemove is > 0 only if we are trying to delete base rule(s).
-            if (toRemove.size() > 0) {
-                removedRules.addAll(switchManager.deleteRuleWithCookie(dpid, toRemove));
-                if (deleteAction == DeleteRulesAction.REMOVE_ADD) {
-                    switchManager.installDefaultRules(dpid);
-                }
-            } else {
+            // The cases when we delete the default rules.
+            if (EnumSet.of(DeleteRulesAction.DROP_ALL, DeleteRulesAction.DROP_ALL_ADD_DEFAULTS,
```

Comment:
This will create a new object, the same object, every function call.  Unless the compiler is smart enough to detect and optimize.  Suggest to make this a private static final and re-use in the method.